### PR TITLE
Used hardhat for testing

### PR DESCRIPTION
**Pull request by tomaszslabon**
This PR modifies tests so that they are run on `hardhat`.

### CODE REVIEW

File: contracts/SortitionPoolFactory.sol [hunk 1]
```diff
@@ -6,21 +6,25 @@ import "./api/IStaking.sol";
...
+    address sortitionPool = address(
+      new SortitionPool(
+        stakingContract,
+        minimumStake,
+        poolWeightDivisor,
+        msg.sender
+      )
+    );
+
+    emit SortitionPoolCreated(sortitionPool);
```
Comment (tomaszslabon) [inline contracts/SortitionPoolFactory.sol:27]: Had to emit event to get the address of the new `sortition pools` in tests.

File: package.json [hunk 2]
```diff
@@ -11,10 +11,11 @@
     "migrations/scripts"
   ],
   "scripts": {
+    "build": "hardhat compile",
     "ganache": "echo 'GANACHE IS NO LONGER USED; use buidler-vm script instead'",
     "buidler-vm": "buidler node",
     "buidler-vm:log": "buidler node | grep '^    \\|console.log'",
```
Comment (lukasz-zimnoch) [inline package.json:17]: Let's remove those commands if we got rid of buidler.

File: package.json [hunk 3]
```diff
@@ -11,10 +11,11 @@
...
-    "test": "NODE_ENV=test truffle test",
```
Comment (lukasz-zimnoch) [inline package.json:17]: Are we sure `NODE_ENV` is not used in tests anymore?

File: test/factoryTest.js [hunk 4]
```diff
@@ -1,64 +1,75 @@
...
+
+    const SortitionPoolFactory = await ethers.getContractFactory(
+      "SortitionPoolFactory",
+    )
+    factory = await SortitionPoolFactory.deploy()
+    await factory.deployed()
   })
 
-  describe("createSortitionPool()", async () => {
+  describe("SortitionPoolFactory", () => {
```
Comment (lukasz-zimnoch) [inline test/factoryTest.js:32]: According to the convention we use in other projects, shouldn't it be `createSortitionPool`? We are in the `factoryTest.js` file and we describe `SortitionPoolFactory` at the top level.

File: test/utils.js [hunk 5]
```diff
@@ -1,16 +1,20 @@
-// eslint-disable-next-line camelcase
-async function deploySystem(deploy_list) {
-  const deployed = {} // name: contract object
-  const linkable = {} // name: linkable address
+// FIXME Retrieves past events. This is a workaround for a known issue described
+// FIXME here: https://github.com/nomiclabs/hardhat/pull/1163
+// FIXME The preferred way of getting events would be using listners:
+// FIXME https://docs.ethers.io/v5/api/contract/contract/#Contract--events
```
Comment (lukasz-zimnoch) [inline test/utils.js:4]: Maybe just:
```
// FIXME: Abcdef 
//        ghijklm
```

File: tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (lukasz-zimnoch) [inline tsconfig.json:1]: Is it needed? We are not using TypeScript in sortition pool tests. Unless we plan to do so eventually.

File: contracts/SortitionPoolFactory.sol (hunk 1 above)
Comment (pdyraga) [reply contracts/SortitionPoolFactory.sol:27]: This is fine. Could also help in any web3-related code.

File: package.json (hunk 2 above)
Comment (tomaszslabon) [reply package.json]: Done in https://github.com/keep-network/sortition-pools/pull/127/commits/4ba45853abf00e3486568229a9bbddac8475f0bb.

File: test/factoryTest.js (hunk 4 above)
Comment (tomaszslabon) [reply test/factoryTest.js]: Yes, fixed in https://github.com/keep-network/sortition-pools/pull/127/commits/4ba45853abf00e3486568229a9bbddac8475f0bb.

File: test/utils.js (hunk 5 above)
Comment (tomaszslabon) [reply test/utils.js]: Fixed in https://github.com/keep-network/sortition-pools/pull/127/commits/4ba45853abf00e3486568229a9bbddac8475f0bb.

File: tsconfig.json (hunk 6 above)
Comment (tomaszslabon) [reply tsconfig.json]: Removed it in https://github.com/keep-network/sortition-pools/pull/127/commits/4ba45853abf00e3486568229a9bbddac8475f0bb.

**Review by lukasz-zimnoch — DISMISSED** [review]
No more comments. Leaving final approval for @pdyraga

**Comment by pdyraga** [comment]
Taking a look right now!

File: package.json (hunk 3 above)
Comment (tomaszslabon) [reply package.json:17]: It seems `NODE_ENV` is not needed for hardhat to run.

Comment (lukasz-zimnoch) [reply package.json:17]: Ok fine

**Comment by tomaszslabon** [comment]
@pdyraga 
@lukasz-zimnoch 

I've noticed there are two files with tests written in solidity that are not executed now: `TestBranch.sol` `TestDynamicArray.sol`. They were executed on the `main` branch.

We will have to replace them with some tests that run on hardhat.

File: test/factoryTest.js [hunk 7]
```diff
@@ -1,64 +1,75 @@
-const SortitionPoolFactory = artifacts.require(
-  "./contracts/SortitionPoolFactory.sol",
-)
-const SortitionPool = artifacts.require("./contracts/SortitionPool.sol")
-const StakingContractStub = artifacts.require("StakingContractStub.sol")
+const { helpers, ethers } = require("hardhat")
+const { expect } = require("chai")
+const utils = require("./utils")
 
-const { mineBlocks } = require("./mineBlocks")
```
Comment (pdyraga) [inline test/factoryTest.js:7]: We still have `test/mineBlocks.js`. Is it needed?

File: test/sortitionPoolTest.js [hunk 8]
```diff
@@ -1,156 +1,181 @@
...
+      })
     })
 
-    it("returns group of expected size if less operators are registered", async () => {
-      await staking.setStake(alice, 2000)
-      await pool.joinPool(alice)
-
-      const group = await pool.selectGroup(5, seed, {
-        from: owner,
+    context("reverts when there are no operators in pool", async () => {
```
Comment (pdyraga) [inline test/sortitionPoolTest.js:77]:
```
context("when there are no operators in the pool", async () => {
   it("should revert",
```

File: test/sortitionTreeTest.js [hunk 9]
```diff
@@ -1,218 +1,275 @@
...
+        ).to.be.revertedWith("Operator is not registered in the pool")
+      })
     })
 
-    it("does not remove ID numbers", async () => {
-      await sortition.publicInsertOperator(alice, 0x1234)
-      await sortition.publicRemoveOperator(alice)
+    context("when operator is registered", () => {
+      beforeEach(async () => {
+        await sortition.publicInsertOperator(alice.address, 0x1234)
```
Comment (pdyraga) [inline test/sortitionTreeTest.js:160]: It's already done in `beforeEach`

File: test/sortitionTreeTest.js [hunk 10]
```diff
@@ -1,218 +1,275 @@
...
     })
   })
 
-  describe("operatorsInPool()", async () => {
-    it("works as expected", async () => {
-      await sortition.publicInsertOperator(alice, 1)
-      const nOperators = await sortition.operatorsInPool.call()
-      assert.equal(nOperators, 1)
+  describe("operatorsInPool", async () => {
+    context("when there is an operator is in the pool", () => {
```
Comment (pdyraga) [inline test/sortitionTreeTest.js:264]: "when the operator is in the pool" ? 😄

File: test/sortitionPoolTest.js (hunk 8 above)
Comment (tomaszslabon) [reply test/sortitionPoolTest.js]: Done in https://github.com/keep-network/sortition-pools/pull/127/commits/1fdcda0aae9ee9065e524e6280f2dc2157e2aaf3.

File: test/sortitionTreeTest.js (hunk 10 above)
Comment (tomaszslabon) [reply test/sortitionTreeTest.js]: Done in https://github.com/keep-network/sortition-pools/pull/127/commits/1fdcda0aae9ee9065e524e6280f2dc2157e2aaf3.

File: test/factoryTest.js (hunk 7 above)
Comment (tomaszslabon) [reply test/factoryTest.js:7]: Done in https://github.com/keep-network/sortition-pools/pull/127/commits/1fdcda0aae9ee9065e524e6280f2dc2157e2aaf3.

**Comment by pdyraga** [comment]
> @pdyraga @lukasz-zimnoch
> 
> I've noticed there are two files with tests written in solidity that are not executed now: `TestBranch.sol` `TestDynamicArray.sol`. They were executed on the `main` branch.
> 
> We will have to replace them with some tests that run on hardhat.

You can use the same trick as we use in random beacon. See:
https://github.com/keep-network/keep-core/blob/16c30709e185136e6fedeeca321a662dedf479c5/solidity/random-beacon/test/AltBn128.test.ts#L84
https://github.com/keep-network/keep-core/blob/16c30709e185136e6fedeeca321a662dedf479c5/solidity/random-beacon/contracts/test/TestAltBn128.sol#L93

**Comment by tomaszslabon** [comment]
> > @pdyraga @lukasz-zimnoch
> > I've noticed there are two files with tests written in solidity that are not executed now: `TestBranch.sol` `TestDynamicArray.sol`. They were executed on the `main` branch.
> > We will have to replace them with some tests that run on hardhat.
> 
> You can use the same trick as we use in random beacon. See: https://github.com/keep-network/keep-core/blob/16c30709e185136e6fedeeca321a662dedf479c5/solidity/random-beacon/test/AltBn128.test.ts#L84 https://github.com/keep-network/keep-core/blob/16c30709e185136e6fedeeca321a662dedf479c5/solidity/random-beacon/contracts/test/TestAltBn128.sol#L93

Done in https://github.com/keep-network/sortition-pools/pull/127/commits/e3aa8021ad7ba41e95b0ccc3e8723484bd016b1a.
Note that in case of `TestBranch.sol`, I removed most of the tests, because they were already tested in `branchTest.js` and would be duplicated.

**Review by pdyraga — APPROVED** [review]

**Pull request merged**